### PR TITLE
Port test application from Python to C++

### PR DESCRIPTION
**Pull request by ubruhin**
As discussed in #37 and #34.

Actually it was pretty easy to port the test application to C++ ;) I first intended to use PyQt, but I guess it would still be a pain for testing, so I decided to go with C++ instead...

Now we no longer have the dependency to PySide or PyQt, and thus the functional tests are much easier to run with all combinations of OS/Python/Qt versions. And CI is *much* faster because PySide no longer needs to be installed :grinning: 

Functional tests are now also run with Qt5 on CI (but only on Linux - Windows and MacOS do not work properly yet, but I think now it should be much easier to fix these issues). We can increase CI test coverage step by step.

What do you think?

### CODE REVIEW

**Comment by ubruhin** [comment]
It seems that the issues on Windows depend on the used Qt version. I updated to Qt 5.11 on AppVeyor, now all tests are passing!

So sooner or later we should test with various Qt versions on CI...

**Comment by ubruhin** [comment]
Yeah, now with the bugfix d45317ccd02f5e13fa044f1a69a66dc36ddb1164 it even works on macOS! :smiley: 

But on Windows the tests seem to be unstable. Sometimes they pass, sometimes not. I'll take a look at that...

File: tests-functionnal/funq-test-app/main.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright: SCLE SFE
+Contributor: Julien Pagès <j.parkouss@gmail.com>
```
[the remark this answers is not in the document]
Comment (parkouss) [reply tests-functionnal/funq-test-app/main.cpp]: Indeed :) Though I am thinking that those lines may be removed one day, I am not sure it makes much sense to have them. (I started from a template given by the company I was working with, but I guess this line is not required).

File: tests-functionnal/funq-test-app/widgets.h [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            _statusLabel->setText(msg);
+        }
+};
+
+class ClickDialog : public SimpleDialog {
+        Q_OBJECT
+    public:
+        ClickDialog(QLabel* statusLabel, QWidget* parent) :
+            SimpleDialog(statusLabel, parent) {
+            QPushButton* btn = new QPushButton("click me");
```
Comment (parkouss) [inline tests-functionnal/funq-test-app/widgets.h]: Indeed, you may set the parent of the created widgets so they won't leak.

File: tests-functionnal/funq-test-app/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ubruhin) [reply tests-functionnal/funq-test-app/main.cpp]: I replaced this line with that one:

    Contributors: https://github.com/parkouss/funq/graphs/contributors

Is this okay?

File: tests-functionnal/funq-test-app/widgets.h (hunk 2 above)
Comment (ubruhin) [reply tests-functionnal/funq-test-app/widgets.h]: As described [here](https://doc.qt.io/qt-5/layout.html#tips-for-using-layouts), the layout reparents the contained widgets so there shouldn't be a memory leak. I also run valgrind, the only memory leaks it reported were `QTableWidgetItem` objects, but I don't know why. [Qt documentation](https://doc.qt.io/qt-5/qtablewidget.html#setItem) says that `QTableWidget` takes ownership over `QTableWidgetItem`...

**Comment by ubruhin** [comment]
For your information, I removed the commit which fixes tests on MacOS because it makes Windows tests failing completely :sob: It seems to be an annoying problem with sockets. On macOS, one can call `connect()` only once, afterwards the socket object is in a strange state and doesn't work anymore. So I created a new socket object for every attempt, but that fails on Windows, because the port stays busy or something like that.

I will open a new PR once I have a solution for this issue ;) It's not really related to this PR anyway.

**Pull request merged**